### PR TITLE
refactor(iroh-net): More renaming of endpoint to direct address

### PR DESCRIPTION
**Pull request by flub**
## Description

The name "endpoint" is overloaded and we decided a while ago to not
use it for things other than `iroh_net::Endpoint`.  This further
renames more places that were called endpoint with the new name we
already introduced for them: direct address.  The direct address
naming is already used consistently on public APIs, this is only for
internal consistency.

## Breaking Changes

metric changed: `magicsock::metrics::Metrics::update_endpoints` -> `Metrics::update_direct_addrs`.

## Change checklist

- [x] Self-review.
- [x] Documentation updates following the [style guide](https://rust-lang.github.io/rfcs/1574-more-api-documentation-conventions.html#appendix-a-full-conventions-text), if relevant.
- [x] Tests if relevant.
- [x] All breaking changes documented.

### CODE REVIEW

File: iroh-net/src/magicsock.rs [hunk 1]
```diff
@@ -394,9 +394,12 @@ impl MagicSock {
...
     pub(super) fn update_direct_addresses(&self, eps: Vec<DirectAddr>) {
-        let updated = self.endpoints.update(DiscoveredEndpoints::new(eps)).is_ok();
+        let updated = self
+            .direct_addrs
+            .update(DiscoveredDirectAddrs::new(eps))
+            .is_ok();
         if updated {
-            let eps = self.endpoints.read();
+            let eps = self.direct_addrs.read();
             eps.log_endpoint_change();
```
Comment (matheus23) [inline iroh-net/src/magicsock.rs:403]:
```suggestion
            let eps = self.direct_addrs.read();
            eps.log_direct_addrs_change();
```

File: iroh-net/src/magicsock.rs [hunk 2]
```diff
@@ -1236,15 +1239,15 @@ impl MagicSock {
...
+        self.direct_addr_update_state.schedule_run(why);
     }
 
     /// Publishes our address to a discovery service, if configured.
     ///
     /// Called whenever our addresses or home relay node changes.
     fn publish_my_addr(&self) {
         if let Some(ref discovery) = self.discovery {
-            let eps = self.endpoints.read();
+            let eps = self.direct_addrs.read();
```
Comment (matheus23) [inline iroh-net/src/magicsock.rs:1250]:
```suggestion
            let direct_addrs = self.direct_addrs.read();
```

File: iroh-net/src/magicsock.rs [hunk 3]
```diff
@@ -1236,15 +1239,15 @@ impl MagicSock {
...
             let relay_url = self.my_relay();
             let direct_addresses = eps.iter().map(|ep| ep.addr).collect();
             let info = AddrInfo {
```
Comment (matheus23) [inline iroh-net/src/magicsock.rs:1253]:
```suggestion
            let direct_addrs = self.direct_addrs.read();
            let relay_url = self.my_relay();
            let addrs = direct_addrs.iter().map(|da| da.addr).collect();
            let info = AddrInfo {
```

File: iroh-net/src/magicsock.rs [hunk 4]
```diff
@@ -2000,39 +2006,39 @@ impl Actor {
...
-                    $already.insert($ipp, $et);
-                    $eps.push(DirectAddr {
+                    $already.insert($ipp, $typ);
+                    $addrs.push(DirectAddr {
                         addr: $ipp,
-                        typ: $et,
+                        typ: $typ,
                     });
                 }
             };
```
Comment (matheus23) [inline iroh-net/src/magicsock.rs:2044]: Not relevant to this PR, but why is this a macro, not a function?

**Review by matheus23 — APPROVED** [review]
Only small nits :v:

File: iroh-net/src/magicsock.rs (hunk 4 above)
Comment (dignifiedquire) [reply iroh-net/src/magicsock.rs:2044]: why not :P

Comment (matheus23) [reply iroh-net/src/magicsock.rs:2044]: I see you half-joking, but I'll still explain myself: It's mostly about complexity. I'd argue macros are more complex and that leads to a bunch of inefficiencies. Tooling gets confused/disabled/slow more often, people are more likely to shy away from that code, etc. Same arguments as in any other case with more complexity, e.g. when there's too many abstraction layers.

Comment (dignifiedquire) [reply iroh-net/src/magicsock.rs:2044]: I never joke ;) 

The answer is when I wrote this originally, it was easier to use a macro, I don't remember why though

File: iroh-net/src/magicsock.rs [hunk 5]
```diff
@@ -229,15 +229,15 @@ pub(crate) struct MagicSock {
     /// Optional discovery service
     discovery: Option<Box<dyn Discovery>>,
 
-    /// Our discovered endpoints
-    endpoints: Watchable<DiscoveredEndpoints>,
+    /// Our discovered direct addresses
```
Comment (divagant-martian) [inline iroh-net/src/magicsock.rs:232]:
```suggestion
    /// Our discovered direct addresses.
```

File: iroh-net/src/magicsock.rs (hunk 1 above)
Comment (divagant-martian) [reply iroh-net/src/magicsock.rs:403]: I would generally avoid non obvious abbreviations. `direct_addrs` is fine by me

File: iroh-net/src/magicsock.rs [hunk 6]
```diff
@@ -1295,25 +1298,24 @@ impl DiscoMessageSource {
...
 /// - if an update is scheduled while another one is running, remember that
 ///   and start a new one when the current one has finished
 #[derive(Debug)]
-struct EndpointUpdateState {
+struct DirectAddrUpdateState {
     /// If running, set to the reason for the currently the update.
     running: sync::watch::Sender<Option<&'static str>>,
-    /// If set, this means we will start a new endpoint update state as soon as the current one
-    /// is finished.
+    /// If set start a new update as soon as the current one is finished.
```
Comment (divagant-martian) [inline iroh-net/src/magicsock.rs:1311]:
```suggestion
    /// If set, start a new update as soon as the current one is finished.
```

File: iroh-net/src/magicsock.rs [hunk 7]
```diff
@@ -2461,59 +2467,63 @@ fn bind(port: u16) -> Result<(UdpConn, Option<UdpConn>)> {
     Ok((pconn4, pconn6))
 }
 
+/// The discovered direct addresses of this [`MagicSock`].
+///
+/// These are all the [`DirectAddr`]s that this [`MagicSock`] is aware of exist for itself.
```
Comment (divagant-martian) [inline iroh-net/src/magicsock.rs:2472]:
```suggestion
/// These are all the [`DirectAddr`]s that this [`MagicSock`] is aware of for itself.
```

another alternative: `These are all the addresses that this endpoint knows exist for itself.`

not sure if a typo when changing the idea mid-write, which happens to all, or intended. However, as is, this is grammatically not correct

**Review by divagant-martian — APPROVED** [review]
minor things. LGTM and I find it easier to read

File: iroh-net/src/magicsock.rs (hunk 4 above)
Comment (flub) [reply iroh-net/src/magicsock.rs:2044]: yeah, I share @matheus23's opinion here.  I tried to re-write this once as a function and it was impossible due to borrowing rules.  It probably is still possible with a bit more refactoring though, I just recall abandoning it for some reason.

**Pull request merged**